docs/amber.md

```diff
@@ -0,0 +1,138 @@
+another engine), calling into the parsing code to generate a test script and
+then passing that script into the script execution component.
+
+We require the application to configure the execution engine. This allows the
+application to handle loading function pointers, doing configuration and other
+setups as required. There are no hardcoded assumptions in amber as to how you're
+setting up the API to be tested.
+
+This engine configuration is done through the `VulkanEngineConfig` in
+`amber/amber_vulkan.h` and the `DawnEngineConfig` in `amber/amber_dawn.h`.
+
+The sample application does this configuration through the config_helper
+classes. `samples/config_helper_vulkan.*` and `samples/config_helper_dawn.*`.
+For the CTS, the Vulkan engine is already configured and we just set the
+`VulkanEngineConfig` as needed.
+
+Accessing amber itself is done through the `Amber` class in `amber/amber.h`. We
+require the application to parse and execute a script as two separate steps.
+This separation allows for the scripts to be retrieved after a parse command
+and then pre-compiled and passed into the execution. Passing the compiled
+scripts is optional.
+
+### Delegate
+A delegate object can be provided to Amber, `nullptr` can be provided for the
+delegate. If the delegate is provided and the `LogGraphicsCalls` method returns
+`true` then the vulkan API wrappers will call `Log` for each call into Vulkan.
+If the `LogGraphicsCallsTime` also returns `true` then timings for those
+Vulkan calls will also be recorded. The timestamps are retrieved from the
+`GetTimestampNS` callback.
+
+### Buffer Extractions
+Amber can be instructed to retrieve the contents of buffers when execution is
+complete. This is done through the `extractions` list in the amber `Options`
+structure. You must set the `buffer_name` for each of the buffers you want to
+extract. When amber completes it will fill out the `width`, `height` and set
+of `Value` objects for that buffer.
+
+### Execution
+There are two methods to execute a parsed script `Execute` and
+`ExecuteWithShaderData`. They both accept the `Recipe` and `Options`, the
+`ExecuteWithShaderData` also accepts a map of shader name to data. The data
+is the compiled SPIR-V binary for the shader. This allows you to compile and
+cache the shader if needed.
+
+## Parsing component
+There are currently two parsing backends for Amber.
+[AmberScript](amber_script.md) and [VkScript](vk_script.md). The `AmberScript`
+parser will be used if the first 7 characters of the input buffer are
+`#!amber`, otherwise the `VkScript` parser will be used. The parsers both
+generate a `Script` which is our representation of the script file.
+
+### VkScript
+For VkScript we do a bit of work to make the script match AmberScript. A default
+pipeline is generated and all content in the script is considered part of the
+generated pipeline. We generate names for all of the buffers in the file. The
```

Comment:
"An Amber buffer corresponds to either a buffer resource or an image resource in the backend engine's API.  We generate a unique name for each buffer in the script."

Reply:
(hmm. but see below, where I suggested a different place for that first sentence.)

Reply:
added it below where I think it makes more sense.